### PR TITLE
Terrain Generation

### PR DESCRIPTION
**Pull request by edwardsCam**
use Perlin noise and linear interpolation to smooth terrain.

### CODE REVIEW

File: common/noise/PerlinNoise.cpp [hunk 1]
```diff
@@ -2,6 +2,11 @@
 #include "PerlinNoise.h"
 #include "tbb/tbb.h"
 
+#define wrap_left  ( i==0 ? width-1 : i-1 )
```
Comment (ksundberg) [inline common/noise/PerlinNoise.cpp]: #define is not appropriate for defining constants, and even worse for macros. 


These need to be redone using safer techniques.

File: common/noise/PerlinNoise.cpp [hunk 2]
```diff
@@ -259,4 +263,31 @@ double PerlinNoise::noise(const vector3d& p) const
...
+//This is an interpolating smoothing function. Can be implemented as a substitute, or with Perlin function for better smoothing.
+void smooth(std::shared_ptr<matrix2d> noiseMap)
+{
+  int width = (*noiseMap).size();
+  int height = (*noiseMap)[0].size();
+  for (int i = 0; i < width; i++)
+  {
+    for (int j = 0; j < height; j++)
+	{
+	  double new = 0.0;
```
Comment (ksundberg) [inline common/noise/PerlinNoise.cpp]: new is a keyword.  I don't know how this compiles for you

File: common/noise/PerlinNoise.cpp [hunk 3]
```diff
@@ -259,4 +258,31 @@ double PerlinNoise::noise(const vector3d& p) const
   sum += knot(fi + 1, fj + 1, fk + 1, v);
 
   return sum;
+}
+
+//This is an interpolating smoothing function. Can be implemented as a substitute, or with Perlin function for better smoothing.
+void smooth(std::shared_ptr<matrix2d> noiseMap)
+{
+  int width = (*noiseMap).size();
```
Comment (ostlerc) [inline common/noise/PerlinNoise.cpp]: why not use the arrow operator? (ie noiseMap->size())

**Comment by edwardsCam** [comment]
Thanks for the comments, I agree with all.

The variable 'p' was being declared twice, once as a PerlinNoise and once as a double.

**Pull request merged**